Eslint plugin jest

jkimbo opened the pull request:
**Summary**

Importing [`eslint-plugin-jest`](https://github.com/jkimbo/eslint-plugin-jest).

I've tried to match the package.json and file conventions that the other packages have to the best of my ability!

cpojer commented:
This is great! I'll check this out soon when everything around me gets a tiny little bit more quiet. Sorry I can't get to it immediately.

orta commented on packages/eslint-plugin-jest/src/rules/no-identical-title.js:
```diff
@@ -0,0 +1,105 @@
+  };
+}
+
+function handlTestCaseTitles(context, titles, node, title) {
```
`handlTestCaseTitles` -> `handleTestCaseTitles` 👍 (and same-ish typo below)

jkimbo replied:
Good catch! Fixed.

jkimbo commented:
@thymikee sure I can give it a go. Not sure what is necessary regarding legal/copyright info though?

thymikee commented:
Adding the text below to source and test files should do the work:
```
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 * @flow
 */
```

cpojer commented:
ping, is the flow typing still happening? :)

jkimbo commented:
@thymikee @cpojer add flow types and copyright notices

SimenB commented on packages/eslint-plugin-jest/README.md:
````diff
@@ -0,0 +1,46 @@
+## Installation
+
+```
+$ yarn install eslint eslint-plugin-jest --dev
````
`yarn add`

thymikee replied:
yarn install is available in v18

SimenB commented on packages/eslint-plugin-jest/README.md:
````diff
@@ -0,0 +1,46 @@
+```json
+{
+  "rules": {
+    "jest/no-exclusive-tests": 2,
````
use `"error"` instead of `2`?

SimenB commented on packages/eslint-plugin-jest/package.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "eslint-plugin-jest",
+  "version": "17.0.0",
```
18?

SimenB commented:
Should there be a recommended export as well? Can be added later of course.

thymikee commented on packages/eslint-plugin-jest/package.json:
```diff
@@ -0,0 +1,24 @@
+  "peerDependencies": {
+    "eslint": ">=3.6"
+  },
+  "license": "ISC"
```
it should be `"license": "BSD-3-Clause"`

thymikee commented on packages/eslint-plugin-jest/src/index.js:
```diff
@@ -0,0 +1,16 @@
+/* eslint-disable global-require */
```
Can you move this below copyright?

thymikee commented on packages/eslint-plugin-jest/src/rules/__tests__/no-exclusive-tests-test.js:
```diff
@@ -0,0 +1,65 @@
+ *
+ * @flow
+ */
+/* eslint-disable sort-keys */
```
newlines above and below `/* eslint-disable sort-keys */`

thymikee commented on packages/eslint-plugin-jest/src/rules/types.js:
```diff
@@ -0,0 +1,28 @@
+ * @flow
+ */
+export type EslintContext = {
+  report: (message: any) => void
```
You're passing `(message: any) => void` but throughout the code it looks like it should be:
`({message: string, node: EslintNode}) => void`, right?

jkimbo replied:
I've re-written the type definitions to be more accurate (although not complete) to the actual ast types.

thymikee commented:
@jkimbo thank you for the updates!

jkimbo commented:
Addressed all the code review comments. 

@SimenB I thought about adding an recommended export but since at the moment there are only 2 rules I didn't think it would be that useful. Should be added in the future though.

thymikee commented:
cc: @cpojer

cpojer commented:
Wow this is really solid! Nice work. I have one more lint rule to add, so I'll do that in a follow-up :)

cpojer commented:
Small follow-up here: https://github.com/facebook/jest/pull/2509

cpojer commented:
Actually, @jkimbo would you mind adding the lint rule I was thinking of? I think we should have one that errors when using `xit` and `xdescribe` that people can optionally enable. At Facebook it's really bad when people disable tests (while they are working on something) and then potentially commit disabled tests (similar to the rule for `test.only`, really).

jkimbo commented:
@cpojer sure! I'll give it a go over the weekend.

jkimbo commented:
@cpojer https://github.com/facebook/jest/pull/2528
